src/kontain.c

```diff
@@ -26,246 +26,33 @@
-      access->major_present = 1;
-      access->minor = dev->minor;
-      access->minor_present = 1;
-      access->access = strdup ("rwm");
-
-      *devp = dev;
-      *accessp = access;
-    }
-  else
-    {
-      ret = errno;
-    }
-  return ret;
-}
-
-int
-add_kontain_device (libcrun_container_t *container, const char *device)
-{
-  runtime_spec_schema_config_schema *container_def = container->container_def;
-  runtime_spec_schema_config_linux *linux = container_def->linux;
-  runtime_spec_schema_defs_linux_device *dev;
-  runtime_spec_schema_defs_linux_device_cgroup *access;
-  int ret = ENODEV;
-
-  ret = build_kontain_device (device, &dev, &access);
-  if (ret != 0)
-    {
-      return ret;
-    }
-
-  // No resources, make an empty one of those first
-  if (linux->resources == NULL)
-    {
-      linux->resources = calloc (1, sizeof (*linux->resources));
-      if (linux->resources == NULL)
-        {
-          return ENOMEM;
-        }
-    }
-
-  // Grow devices array
-  size_t new_devices_len = linux->devices_len + 1;
-  runtime_spec_schema_defs_linux_device **new_devices = realloc (linux->devices, new_devices_len * sizeof (runtime_spec_schema_defs_linux_device **));
-  if (new_devices == NULL)
-    {
-      return ENOMEM;
-    }
-  new_devices[linux->devices_len] = dev;
-  linux->devices = new_devices;
-  linux->devices_len = new_devices_len;
-
-  // Grow access array
-  size_t new_res_devices_len = linux->resources->devices_len + 1;
-  runtime_spec_schema_defs_linux_device_cgroup **new_res_devices = realloc (linux->resources->devices, new_res_devices_len * sizeof (runtime_spec_schema_defs_linux_device_cgroup));
-  if (new_res_devices == NULL)
-    {
-      return ENOMEM;
-    }
-  new_res_devices[linux->resources->devices_len] = access;
-  linux->resources->devices = new_res_devices;
-  linux->resources->devices_len = new_res_devices_len;
-
-  return 0;
-}
-
-int
-add_kontain_devices (libcrun_container_t *container)
-{
-  int ret = 0;
-  struct stat statb;
-
-  if (stat ("/dev/kvm", &statb) == 0)
-    {
-      ret = add_kontain_device (container, "/dev/kvm");
-      if (ret != 0)
-        return ret;
-    }
-  if (stat ("/dev/kkm", &statb) == 0)
-    {
-      ret = add_kontain_device (container, "/dev/kkm");
-      if (ret != 0)
-        return ret;
-    }
-  return 0;
-}
-
-int
-add_kontain_config (libcrun_container_t *container)
+/*
+ * Get path to km (originally hardcoded as /opt/kontain/bin/km)
+ *
+ * Assume km it located right next to krun
+ *
+ */
+char *
+get_km_bin_path (char *krun_path)
 {
-  int ret;
-  const char *privileged = find_annotation (container, "io.podman.annotations.privileged");
+  char *km_bin_path = strdup (krun_path);
 
-  ret = add_kontain_bind_mounts (container, privileged);
-  if (ret != 0)
-    {
-      return ret;
-    }
+  /* Depending on the fact that krun(or krun-trigger-label) strings are shorter then km
+   * so we have enough space
+   */
+  char *dir = dirname (km_bin_path);
```

Comment:
This is not a good thing to do.  The dirname(3) man page does not guarantee what piece of memory it is returning a pointer to.
In other words it may not be returning a pointer into the string you passed it.  Could you use strrchr() to find the last slash in the string?
I'm a little surprised the compiler didn't complain about dir being set but unused.

Reply:
I have changed to using strrchr() per your suggestion